cli/tests/test_parallelize_command.rs

```diff
@@ -0,0 +1,181 @@
+    @  b911505e443e 6
+    ◉  2e00cb15c7b6 5
+    ◉  9df3c87db1a2 4
+    ◉  9f5b59fa4622 3
+    ◉  d826910d21fb 2
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    test_env.jj_cmd_ok(&workspace_path, &["parallelize", "dc0::"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  6c7b60a45eb6 6
+    │ ◉  296f48966777 5
+    ├─╯
+    │ ◉  524062469789 4
+    ├─╯
+    │ ◉  a9334ecaa379 3
+    ├─╯
+    │ ◉  3a7b37ebe843 2
+    ├─╯
+    │ ◉  761e67df44b7 1
+    ├─╯
+    ◉  000000000000
+    "###);
+}
+
+// Only the head commit has descendants.
+#[test]
+fn test_parallelize_with_descendants_simple() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+
+    for n in 1..6 {
+        test_env.jj_cmd_ok(&workspace_path, &["commit", &format!("-m{n}")]);
+    }
+    test_env.jj_cmd_ok(&workspace_path, &["describe", "-m=6"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  b911505e443e 6
+    ◉  2e00cb15c7b6 5
+    ◉  9df3c87db1a2 4
+    ◉  9f5b59fa4622 3
+    ◉  d826910d21fb 2
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    test_env.jj_cmd_ok(&workspace_path, &["parallelize", "dc0::9df"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  f28f986c7134 6
+    ◉        21e9963ac5ff 5
+    ├─┬─┬─╮
+    │ │ │ ◉  524062469789 4
+    │ │ ◉ │  a9334ecaa379 3
+    │ │ ├─╯
+    │ ◉ │  3a7b37ebe843 2
+    │ ├─╯
+    ◉ │  761e67df44b7 1
+    ├─╯
+    ◉  000000000000
+    "###);
+}
+
+// More than one commit being parallelized has descendants, but only the
+// descendants of the head commit will be rebased onto the target commits.
+#[test]
+fn test_parallelize_with_descendants_complex() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["init", "repo", "--git"]);
+    let workspace_path = test_env.env_root().join("repo");
+
+    for n in 1..6 {
+        test_env.jj_cmd_ok(&workspace_path, &["commit", &format!("-m{n}")]);
+    }
+    test_env.jj_cmd_ok(&workspace_path, &["new", "d82", "-m=child-of-2"]);
+    test_env.jj_cmd_ok(&workspace_path, &["new", "2e0", "-m=6"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  d27ee705f7a9 6
+    ◉  2e00cb15c7b6 5
+    ◉  9df3c87db1a2 4
+    ◉  9f5b59fa4622 3
+    │ ◉  60c93e41d09c child-of-2
+    ├─╯
+    ◉  d826910d21fb 2
+    ◉  dc0e5d6135ce 1
+    ◉  000000000000
+    "###);
+
+    test_env.jj_cmd_ok(&workspace_path, &["parallelize", "dc0::9df"]);
+    insta::assert_snapshot!(get_log_output(&test_env, &workspace_path), @r###"
+    @  9f1bec0d6c46 6
+    ◉        7dd2f5648395 5
+    ├─┬─┬─╮
+    │ │ │ ◉  b8f977c12383 4
+    │ │ ◉ │  7be8374575b9 3
+    │ │ ├─╯
+    ◉ │ │  2bfe3fe3e472 1
+    ├───╯
+    │ │ ◉  bf7eef1f1dc6 child-of-2
+    │ ├─╯
```

Comment:
I feel like this commit should have both commit 1 and commit 2 as parents. It depended on both before the rebase.

Good test case, btw. I had not thought of this case before.

Reply:
Oh interesting, I think you're right. I'll report back once I have it fixed.

Reply:
I think this is fixed in the WIP commit I just pushed. The "child-of-2" commit is now a child of 1 and 2 after `jj parallelize` finishes.